[PR] Fix CI API tests failing

LuchoTurtle opened the pull request:
closes #297 
closes #315

LuchoTurtle commented:
This is super weird. The server **is running** and I've tested the tests locally and they all pass. But it seems that the tests are not reaching the server running on `localhost`. 🤔

LuchoTurtle commented:
I've looked everywhere but none of the issues help the case. The connection to `localhost` seems to be not be reachable: 
- https://github.com/lazy-actions/slatify/issues/12
- https://github.com/electron-userland/spectron/issues/862
- https://github.com/sindresorhus/got/issues/1572
- https://github.com/Seeed-Studio/node-red-contrib-ml/issues/4
- https://github.com/mswjs/msw/issues/692

Again, running these tests on my computer works. And I'm running `mix phx.server` and running the tests against the live running server. However this seems to not be the case within github actions (for some reason, it has *stopped establishing a connection* on the server running within the service of the action pipeline).

nelsonic commented:
@LuchoTurtle this is looking promising. 👌 
Assign when ready, then take a look at upgrading the project: https://github.com/dwyl/mvp/issues/322 🙏

LuchoTurtle commented:
@nelsonic should be in a reviewable state. 
I refactored the Github Action files, splitting them into two different ones (one for PRs and another for the `main` branch`) to improve readability and separating concerns.

The API definition tests should now correctly execute in the wanted order and there is no need to seed data. Every resource needed to run the tests is made at `Hoppscotch` level.

Also refactored the `review-app.md` file a bit to include these new changes and fixed some typos I found along the way.

nelsonic commented on .github/workflows/main_ci.yml:
```diff
@@ -3,8 +3,6 @@ name: Elixir CI
 on:
```
Renaming this file will break the `build` badge in the `README.md` 💔 
Also, I don't understand the _need_ to rename it ... 🤷‍♂️ 
What am I missing? 💭

LuchoTurtle replied:
I renamed the pipeline job to go in tandem with the other changes pertaining to readability. I think `Elixir CI` is too broad. Having it changed to `PR CI` is more discernable, as we know the `PR_ci.yml` workflow is running and pertains to the PR, not a merge to the `main` branch (which occurs on another file).

nelsonic replied:
Renamed to `pr.yml` clear enough. no need for mixed case in the file name. 👌

nelsonic commented on lib/api/fly_dev.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "host": "https://mvp-pr-314.fly.dev",
```
does this hard-coding of the `host` mean that this will only work while the `314` PR is open? 💭 
What happens to the _next_ PR that doesn't have this ID?

LuchoTurtle replied:
*dang it*, I thought I had gotten everything covered. I knew one last step was missing. You are right, this was a temporary fix just to see if the runs were being correctly executed on my local machine back-to-back.

Please assign the PR back to me, I will address this. Sorry about this, completely forgot this last step.

nelsonic replied:
Sure thing. re-assigned to you. Please review your own PRs before assigning them to ensure you have checked everything. 🙏

nelsonic commented:
@LuchoTurtle do you think we could remove the line:
https://github.com/dwyl/mvp/blob/f96d85429574a77a3bc00080bac49e24bc6a85e4/.github/workflows/pr.yml#L73

So that the deployment of the `review_app` is not dependent (Waiting for) the `build` ... 💭 
Is there an _reason_ that these two jobs need to be run in series?
Could they be run in parallel so that we aren't waiting?

If the `build & test` fails we just focus on that. 
But I feel like _waiting_ for both jobs to run in series while we're debugging this is mega tedious. 

I feel like we're burning a _lot_ of time waiting for the CI to run to debug this ... ⏳ 
Reminds me of: 

![image](https://user-images.githubusercontent.com/194400/221819294-afc69c18-3ce0-4143-8812-8a757705b02c.png)

LuchoTurtle commented:
The `json` file seems to already be switching properly.
But for *some odd reason*, the first test fails

```sh
Run hopp test -e ./lib/api/fly_dev.json ./lib/api/MVP.json

Running: MVP/Tags/Create tag
 POST  https://mvp-pr-31[4](https://github.com/dwyl/mvp/actions/runs/4291630310/jobs/7477259821#step:5:5).fly.dev/api/tags ERROR
⚠ Error running request.
⚠ Error running test-script.
```

While the other ones all succeed. This didn't happen prior to changing the `json` file, which doesn't make sense. The error is not shown either *and* I can't reproduce this on my `localhost` :/

nelsonic commented:
OK. looks like you've figured out the "hard" part of injecting the review app URL into the `.json` file ... 🎉 
there doesn't _appear_ to be a way of running `hopp` with a "verbose" flag/option ... https://docs.hoppscotch.io/cli 🤷‍♂️

LuchoTurtle commented:
That appears to have fixed it.
Going to change the `build-review-apps.md` file before submitting for review.

LuchoTurtle commented:
Nevermind. I don't know why Github is showing it was unsuccessful when the https://github.com/dwyl/mvp/actions/runs/4291836918 run was ...

<img width="939" alt="image" src="https://user-images.githubusercontent.com/17494745/221826273-eb8b8c06-7db5-4c13-8249-a367dc590d67.png">

LuchoTurtle commented:
Ok, that should do it. Over to you, @nelsonic

nelsonic reviewed (approved):
@LuchoTurtle looks good. 👌
Really hope this is more robust now. 🤞🏼
Thanks. ☀️
